### ISSUE TITLE
STM that contains only `get` operations can see partial state

### ISSUE THREAD

**Issue by adamgfraser** [zio/zio#1587]
https://circleci.com/gh/zio/zio/21901?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

```
[info]         fold:
[error]           x is atomic
[error]  30 != 0; 30 != 1000 (TArraySpec.scala:99)
```

**Comment by ghostdogpr** [comment]
@LGLO any idea about this one?

**Comment by adamgfraser** [comment]
Just saw another issue here. This time a different method. Is there something we changed that could have impacted this?

https://circleci.com/gh/zio/zio/22084?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

[info]         map:
[error]           x creates new array atomically
[error]  19 -> 3 != 19 -> 19; 19 -> 3 != 3 -> 3 (TArraySpec.scala:137)

**Comment by dkarlinsky** [comment]
@adamgfraser I think I managed to create a test that (pretty much) consistently reproduces the failure.
[Here the code](https://github.com/dkarlinsky/zio/blob/3e4edac589ae5d88bfd66bca28da3209797feff7/core-tests/jvm/src/test/scala/zio/stm/TArrayFlakySpec.scala)
It basically runs `TArraySpec.foldAtomic` up to 5000 times, with batches of 50 in parallel, until the assertion of `equalTo(0) or equalTo(N)` fails.
The parallelism seems to be important - before I introduced it I had no luck getting a failure.
I've also run this test on the [commit](https://github.com/zio/zio/commit/934967ad0b2fd32ba535fb3548df2543d042bb78) that introduced `TArray` and `TArraySpec` and it failed there as well.

**Comment by LGLO** [comment]
Thank you @dkarlinsky . I added this code and I feel guilty - I'll take a look this weekend! Hundredfold thanks for providing code!

**Comment by dkarlinsky** [comment]
Sure :) I was hoping to be able to figure out what was going on myself, but after digging through STM internals for a while and feeling out my depth,    I figured I'd provide the test for you or whoever takes this issue. Hope it helps :) 
Very interested to see what the bug ends up being.

**Comment by LGLO** [comment]
I was able to reproduce this bug without TArray but only TRefs:
```
  def test = {
    for {
      r0 <- TRef.makeCommit(0)
      r1 <- TRef.makeCommit(0)
      r2 <- TRef.makeCommit(0)
      r3 <- TRef.makeCommit(0)
      r4 <- TRef.makeCommit(0)
      sum1Fiber <- r0.get.flatMap(v0 => r1.get.flatMap(v1 => r2.get.flatMap(v2 => r3.get.flatMap(v3 => r4.get.map(v4 => v0+v1+v2+v3+v4))))).commit.fork
      _         <- r0.update(_ + 1).flatMap(_ => r1.update(_ + 1).flatMap(_ => r2.update(_ + 1).flatMap(_ => r3.update(_ + 1).flatMap(_ => r4.update(_ + 1))))).commit
      sum1      <- sum1Fiber.join
    } yield sum1
  }
```
I'm using more 100 000 iterations.

**Comment by LGLO** [comment]
I was able "fix" the problem by changing `r1.get` to `r1.update(identity)`.
This pointed me that there can be something with `get`.
It seems that when `STM` is made only with `get` operations then journal analysis yields `JournalAnalysis.ReadOnly` (obvsly). However in `STM#tryCommit` we don't check if journal is valid in case of `ReadOnly`. I'll open PR with this check and modified version of test from @dkarlinsky.

**Comment by ghostdogpr** [comment]
Fixed by #1613
